File: web/src/domain/interfaces/HabitTrackRepository.ts
import type { HabitTrack } from '../entities/HabitTrack';

/** Port interface for habit track persistence. */
export interface HabitTrackRepository {
  findById(id: string): Promise<HabitTrack | undefined>;
  findActiveByGoalId(goalId: string): Promise<HabitTrack[]>;
  findActiveByUserId(userId: string): Promise<HabitTrack[]>;
  save(track: HabitTrack): Promise<HabitTrack>;
}

File: backend/domain/src/main/kotlin/com/dopashift/domain/repository/HabitTrackRepository.kt
package com.dopashift.domain.repository

import com.dopashift.domain.entity.HabitTrack
import java.util.UUID

/**
 * Port interface for habit track persistence.
 * Implementations live in the infrastructure layer.
 */
interface HabitTrackRepository {
    suspend fun findById(id: UUID): HabitTrack?
    suspend fun findActiveByGoalId(goalId: UUID): List<HabitTrack>
    suspend fun findActiveByUserId(userId: UUID): List<HabitTrack>
    suspend fun save(track: HabitTrack): HabitTrack
    suspend fun deleteByGoalId(goalId: UUID)
    suspend fun reassignToGoal(fromGoalId: UUID, toGoalId: UUID)
}
